packages/core/src/event_dispatch/event_delegation.ts

```diff
@@ -0,0 +1,38 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {ENVIRONMENT_INITIALIZER, Injector} from '../di';
+import {inject} from '../di/injector_compatibility';
+import {Provider} from '../di/interface/provider';
+import {setStashFn} from '../render3/instructions/listener';
+import {initGlobalEventDelegation, sharedStashFunction} from '../event_delegation_utils';
+import {GlobalEventDelegation} from './global_event_delegation';
+
+import {IS_GLOBAL_EVENT_DELEGATION_ENABLED} from '../hydration/tokens';
+
+/**
+ * Returns a set of providers required to setup support for event delegation.
+ */
+export function provideGlobalEventDelegation(): Provider[] {
+  return [
+    {
+      provide: IS_GLOBAL_EVENT_DELEGATION_ENABLED,
+      useValue: true,
+    },
+    {
```

Comment:
We can provide a different implementation of a class here:
```suggestion
    },
    {
      provide: GlobalEventDelegation,
      useClass: GlobalEventDelegationImpl
    },
    {
```

and the implementation may look like this:

```
@Injectable()
export class GlobalEventDelegationImpl extends GlobalEventDelegation {
  eventContract!: EventContract;
  addEvent(el: Element, eventName: string) {
    if (this.eventContract) {
      this.eventContract.addEvent(eventName);
      setJSActionAttribute(el, [eventName]);
      return true;
    }
    return false;
  }
}
```

In this case we don't need to re-assign `addEvent` function on the instance (as we do it in the `packages/core/src/event_delegation_utils.ts` function in this PR).

Reply:
This ended up being really complicated. I didn't want to do the addEvent check in dom_events, so I added provideEventDelegation to @angular/platform-browser. Using a different implementation of DomEventsPlugin didn't work, so I monkey patched the prototype of DomEventsPlugin instead...

Reply:
ok changed to what we discussed offline with a "frameworkPlugin" flag

Reply:
changed again to an overriding implementation of GlobalEventDelegation